src/front/wgsl/mod.rs

```diff
@@ -328,12 +328,50 @@ impl<'a> ExpressionContext<'a, '_, '_> {
         let mut left = parser(lexer, self.reborrow())?;
         while let Some(op) = classifier(lexer.peek().0) {
             let _ = lexer.next();
-            let expression = crate::Expression::Binary {
-                op,
-                left,
-                right: parser(lexer, self.reborrow())?,
-            };
-            left = self.expressions.append(expression);
+            let right = parser(lexer, self.reborrow())?;
+            left = self
+                .expressions
+                .append(crate::Expression::Binary { op, left, right });
+        }
+        Ok(left)
+    }
+
+    fn parse_binary_splat_op(
+        &mut self,
+        lexer: &mut Lexer<'a>,
+        classifier: impl Fn(Token<'a>) -> Option<crate::BinaryOperator>,
+        mut parser: impl FnMut(
+            &mut Lexer<'a>,
+            ExpressionContext<'a, '_, '_>,
+        ) -> Result<Handle<crate::Expression>, Error<'a>>,
+    ) -> Result<Handle<crate::Expression>, Error<'a>> {
+        let mut left = parser(lexer, self.reborrow())?;
+        while let Some(op) = classifier(lexer.peek().0) {
+            let _ = lexer.next();
+            let mut right = parser(lexer, self.reborrow())?;
+            // insert splats, if needed by the non-'*' operations
+            if op != crate::BinaryOperator::Multiply {
```

Comment:
The reason `Multiply` is handled specially here is that it handles splats in the IR, correct? I wonder if it wouldn't ultimately make the IR simpler to make all splats explicit. But either way, that shouldn't be part of this PR.

Reply:
The IR currently follows SPIR-V, where "*" has a special place, because it's always nicely accelerated (i.e. multiply-add instructions, matrix-vector multiplication, etc), and also is sounds mathematically (linear scale). So even if we change the "*" to not be special for splats, it's still going to be special.